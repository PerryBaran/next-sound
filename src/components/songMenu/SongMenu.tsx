"use client"

import { useState } from "react"
import css from "./songMenu.module.css"
import { usePlaylistContext } from "../../context/PlaylistContext"
import { play, addList, dots } from "../../media/icons/index"
import Image from "next/image"

interface Props {
  songs: {
    songName: string
    audio: string
    image?: string
    artistName: string
    albumName: string
  }[]
}

export default function SongMenu({ songs }: Props) {
  const { handleAddToPlaylist } = usePlaylistContext()
  const [visible, setVisible] = useState(false)

  const handleBlur = () => {
    setVisible(false)
  }

  return (
    <div className={css.container} onBlur={handleBlur}>
      {visible ? (
        <div className={css.options}>
          <button onMouseDown={() => handleAddToPlaylist(songs, true, true)}>
            <span className={css["song-menu-image-container"]}>
              <Image src={play} alt="play now" width={9} />
            </span>
            <span>Play Now</span>
          </button>
          <button
            onMouseDown={() => handleAddToPlaylist(songs, true, false)}
            className={css["play-next"]}
          >
            <span className={css["song-menu-image-container"]}>
              <Image src={addList} alt="play now" width={15} />
            </span>
            <span>Play Next</span>
          </button>
          <button onMouseDown={() => handleAddToPlaylist(songs, false, false)}>
            <span className={css["song-menu-image-container"]}>
              <Image src={addList} alt="play now" width={15} />
            </span>
            <span>Add To Queue</span>
          </button>
        </div>
      ) : null}
      <button
        onClick={() => setVisible((prev) => !prev)}
        className={css.button}
      >
        <Image src={dots} alt="..." height={18} width={18} />
      </button>
    </div>
  )
}
