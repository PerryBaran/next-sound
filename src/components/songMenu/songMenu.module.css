.container {
  position: relative;
}

.options {
  position: absolute;
  bottom: 20px;
  right: 0px;
  width: 110px;
  display: flex;
  flex-direction: column;
  border: 1px solid var(--darker-grey);
  overflow: hidden;
  z-index: 1000;
}

.options > button {
  border: none;
  border-bottom: 1px solid var(--darker-grey);
  background-color: var(--almost-white);
  color: var(--darker-grey);
  font-weight: 100;
  cursor: pointer;
  text-align: left;
  display: flex;
  height: 20px;
}

.options > button:hover {
  background-color: var(--main-colour);
  color: var(--almost-white);
}

.options > button:last-child {
  border-bottom: none;
}

.image-container {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 20px;
  height: 100%;
}

.options > button:hover img {
  filter: invert();
}

.play-next img {
  transform: scaleY(-1);
}

.button {
  border: none;
  background: none;
  padding: 5px;
  display: flex;
  align-items: center;
  justify-content: center;
  cursor: pointer;
}

.button > img {
  transition-duration: 250ms;
  transform: rotate(90deg);
}

.button:hover > img {
  transform: rotate(90deg) scale(1.25);
}
