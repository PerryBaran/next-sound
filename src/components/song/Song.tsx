import Image from 'next/image'
import css from './song.module.css'
import { music } from '@/media/icons'
import Link from 'next/link'
import SongMenu from '../songMenu/SongMenu'

interface Props {
  artistName: string
  albumName: string
  albumArt: string
  songName: string
  songAudio: string
}
export default function Song({
  artistName,
  albumName,
  albumArt,
  songName,
  songAudio,
}: Props) {
  return (
    <div className={css.song}>
      <div className={css["song__info-container"]}>
        <Image
          src={albumArt || music}
          alt={`${albumName} cover art`}
          height={100}
          width={100}
        />
        <div className={css["song__info"]}>
          <h2 className={css["song__artist-name"]}>
            <Link href={`/profile/${artistName}`}>{artistName}</Link>
          </h2>
          <h3 className={css["song__song-name"]}>
            {songName}
          </h3>
        </div>
        <SongMenu 
          songs={[{
            songName,
            audio: songAudio,
            image: albumArt,
            artistName,
            albumName
          }]}
        />
      </div>
    </div>
  )
}