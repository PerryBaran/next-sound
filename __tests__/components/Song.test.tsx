import { render, screen } from "@testing-library/react"
import Song from "../../src/components/song/Song"
import * as SongMenu from "../../src/components/songMenu/SongMenu"

describe("Song", () => {
  const props = {
    artistName: "artist-name",
    albumName: "album-name",
    albumArt: "/album-art",
    songName: "song-name",
    songAudio: "song-audio"
  }

  const mockedSongMenu = jest.fn()

  jest.spyOn(SongMenu, "default").mockImplementation((props: any) => {
    mockedSongMenu(props)
    return <div />
  })

  test("renders correctly", () => {
    render(<Song {...props} />)

    expect(screen.getByAltText(`${props.albumName} cover art`)).toHaveAttribute("src", `/_next/image?url=%2F${props.albumArt.substring(
      1,
      props.albumArt.length
    )}&w=256&q=75`)
    expect(screen.getByText(props.artistName)).toHaveAttribute("href", `/profile/${props.artistName}`)
    expect(screen.getByText(props.songName)).toBeInTheDocument()
    expect(mockedSongMenu).toBeCalledTimes(1)
    expect(mockedSongMenu).toBeCalledWith({ songs: [{
      songName: props.songName,
      audio: props.songAudio,
      image: props.albumArt,
      artistName: props.artistName,
      albumName: props.albumName
    }]})
  })

  test("snapshot", () => {
    const { asFragment } = render(<Song {...props} />)

    expect(asFragment()).toMatchSnapshot()
  })
})