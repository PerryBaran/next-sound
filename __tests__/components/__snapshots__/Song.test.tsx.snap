// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Song snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="info-container"
    >
      <img
        alt="album-name cover art"
        data-nimg="1"
        decoding="async"
        height="100"
        loading="lazy"
        src="/_next/image?url=%2Falbum-art&w=256&q=75"
        srcset="/_next/image?url=%2Falbum-art&w=128&q=75 1x, /_next/image?url=%2Falbum-art&w=256&q=75 2x"
        style="color: transparent;"
        width="100"
      />
      <div
        class="info"
      >
        <h2
          class="artist-name"
        >
          <a
            href="/profile/artist-name"
          >
            artist-name
          </a>
        </h2>
        <h3
          class="song-name"
        >
          song-name
        </h3>
      </div>
    </div>
    <div />
  </div>
</DocumentFragment>
`;
